keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,815 @@
+- Updating immutable fields
+- Reject updates to VolumeSnapshot’s VolumeSnapshotSource
+- Reject updates to VolumeSnapshotContent’s VolumeSnapshotContentSource
+- Reject updates to VolumeSnapshotContent’s volume snapshot ref after binding
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Backwards compatibility
+
+There is a backwards compatibility issue involved when tightening the validation on snapshot objects. Since the feature is already in beta we are committed to more backward compatibility guarantee than alpha.
+
+There are two perspectives.
+- client perspective: API requests that previously succeeded now fail (these are okay)
+  - create: invalid things get rejected rather than persisted
+  - update: invalid transitions now get rejected
+- API server perspective for existing persisted data
+  - do not tighten validation (via schema or webhook) in a way that would prevent a no-op update (this must not be violated)
+  - reason: removing a finalizer is done via update, causes problems with delete
+  - making a previously optional field required in the schema blocks update of previously persisted data that omitted the field (unless the update populates the newly required field, or you specify a schema default)
+
+To tackle the backwards compatibility problem, this KEP proposes the following release process.
+
+Begin with validating webhook only enforcement. The webhook will perform the following validation
+- one release with ratcheting validation using the webhook server
+   - webhook is strict on create
+   - webhook is strict on updates where the existing object passes strict validation
+   - webhook is relaxed on updates where the existing object fails strict validation (allows finalizer removal, status update, deletion, etc)
+ - some reconciliation process in volume snapshot controller that ensures invalid data is fixed or removed. This process will be entirely user managed. The webhook and controllers will not take any automatic action to reconcile invalid objects.
+
+Once we are sure no invalid data is persisted, we can switch to CRD schema-enforced validation with validating webhooks for immutability in a subsequent release.
+
+If users do not completely remove their invalid objects before upgrading their CRD definition, it should be possible to downgrade the CRD definition to allow invalid objects to get deleted.
```

Comment:
In many cases, users don't directly upgrade the CRD definition, it's done indirectly by upgrading the cluster. So downgrading the CRD version could mean they have to be able to downgrade the cluster control plane.

Reply:
And is it safe to downgrade a CRD definition if there were objects newly created using the new API version? Do we need to control the storage version so that it's still on v1beta1 for one release? cc @liggitt

Reply:
> is it safe to downgrade a CRD definition if there were objects newly created using the new API version?

It is not safe to remove a version from `spec.versions` if there are objects persisted in etcd using that version.

The API server actually will not allow you to update a CRD to do that, so attempting to apply a back-level CRD in a way that would remove a version that was used as a storage version will not succeed. See https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L132-L148

Reply:
Is it possible to downgrade the control plane then to be able to allow deletion of invalid objects after the newer API version is enabled?  Can we specify a new API version (ie v1 with the stricter validation), but still make the storage version v1beta1?

Reply:
> Is it possible to downgrade the control plane then to be able to allow deletion of invalid objects after the newer API version is enabled?

I'm not quite sure what you mean by this... it should be possible to delete invalid objects via the newer API as well

> Can we specify a new API version (ie v1 with the stricter validation), but still make the storage version v1beta1?

sure

Reply:
I think the concern here was that if we move to a newer API version (v1) that has the oneof and immutability enforced in the CRD schema, then we would not be able to delete existing v1beta1 invalid objects because they would get converted to v1 and fail validation. Does keeping the storage version at v1beta1 allow us to delete invalid objects without needing to rollback anything?

I guess it also mainly just kicks the can down the road for purging these invalid objects. Eventually we should switch the storage version to v1.

Reply:
> I think the concern here was that if we move to a newer API version (v1) that has the oneof and immutability enforced in the CRD schema, then we would not be able to delete existing v1beta1 invalid objects because they would get converted to v1 and fail validation. Does keeping the storage version at v1beta1 allow us to delete invalid objects without needing to rollback anything?

You could send delete requests via v1beta1 and v1, even with schema-invalid data.

The issue is anything in the delete path that requires doing an update, like removing a finalizer. Those updates would fail if the object didn't pass schema validation. Controllers that operate on generic resource types (like the garbage collector controller) will prefer the newest, most stable version, so they would operate against v1 rather than v1beta1. Thinking about this more, it's probably worth an issue for sig-api-machinery to tolerate gc-only updates (ownerReferences, finalizers, managedFields) to custom resource instances for this scenario.

Reply:
Thanks for the explanation, so without any changes in api-machinery, it sounds like the rollback procedure would be something like:

1. New v1 api with v1beta1 storage is causing problems and user wants to rollback
2. Control plane (including CRD + snapshot controller) is rolled back and the v1 version is dropped. Objects still persist though since it's stored as v1beta1
3. User fixes their problems with invalid objects
4. User upgrades the control plane again.
5. In an n+2 release, once all the invalid objects are purged, we can switch the storage version to v1.

Reply:
> New v1 api with v1beta1 storage is causing problems and user wants to rollback

I've lost track of the context, is this assuming the following version config in that release:
* v1 (served=true, storage=false)
* v1beta1 (served=false, storage=true)

keeping storage version at v1beta1 allows rollback, but not serving v1beta1 ensures no new invalid objects are created

If so, the rollout approach you described sounds right

Reply:
Yes, that's mostly what I had in mind. Is it possible to also still serve v1beta1 with ratcheting validation in the webhook to prevent new invalid objects?

Reply:
I added a note with the rollback procedure michelle suggested.

Reply:
I also noted that there will be effectively 3 phases. The first two will be as described before, except in phase 2 we will keep the storage version as v1beta1. In the last release / phase the storage version is changed to v1.

Reply:
Opened https://github.com/kubernetes/kubernetes/issues/94060 to discuss in api-machinery